File: 2024/seven/solution.clj
(ns advent-of-code.y2024.day07
  (:require [clojure.string :as str]))

(defn parse-line [line]
  (let [nums (map #(Long/parseLong %) (re-seq #"\d+" line))]
    {:target (first nums)
     :nums (rest nums)}))

(defn parse-input [input]
  (map parse-line (str/split-lines input)))

(defn check1 [target acc nums]
  (cond
    (empty? nums) (= target acc)
    :else (or (check1 target (* acc (first nums)) (rest nums))
              (check1 target (+ acc (first nums)) (rest nums)))))

(defn check2 [target acc nums]
  (cond
    (> acc target) false
    (empty? nums) (= target acc)
    :else (or (check2 target (Long/parseLong (str acc (first nums))) (rest nums))
              (check2 target (* acc (first nums)) (rest nums))
              (check2 target (+ acc (first nums)) (rest nums)))))

(defn filter-valid [input checker]
  (->> (parse-input input)
       (filter (fn [{:keys [target nums]}]
                 (checker target (first nums) (rest nums))))
       (map :target)))

(defn part-one [input]
  (reduce + (filter-valid input check1)))

(defn part-two [input]
  (reduce + (filter-valid input check2)))

(defn read-input []
  (slurp "input.txt"))

(defn -main []
  (let [input-data (read-input)]
    (println "Part One:" (part-one input-data))
    (println "Part Two:" (part-two input-data))))

(-main)

File: 2024/seven/solution.py
import re

def parse_input(file_path):
    with open(file_path, 'r') as f:
        return [
            (int(nums[0]), list(map(int, nums[1:])))
            for line in f.read().splitlines()
            if (nums := re.findall(r'\d+', line))
        ]

def check1(target, acc, nums):
    if not nums:
        return target == acc
    n = nums[0]
    rest = nums[1:]
    return check1(target, acc * n, rest) or check1(target, acc + n, rest)

def check2(target, acc, nums):
    if acc > target:
        return False
    if not nums:
        return target == acc
    n = nums[0]
    rest = nums[1:]
    return (check2(target, int(f"{acc}{n}"), rest) or
            check2(target, acc * n, rest) or 
            check2(target, acc + n, rest))

def filter_valid(data, checker):
    return (target for target, nums in data if checker(target, nums[0], nums[1:]))

def part_one(data):
    return sum(filter_valid(data, check1))

def part_two(data):
    return sum(filter_valid(data, check2))

if __name__ == "__main__":
    input_file = "input.txt"
    data = parse_input(input_file)
    print("Part One:", part_one(data))
    print("Part Two:", part_two(data))
